src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunXmlRootProvider.cs

```diff
@@ -0,0 +1,217 @@
+        private readonly Stream _documentStream;
+        private readonly ManifestResource _resource;
+        private readonly ModuleDesc _owningModule;
+        private readonly string _xmlDocumentLocation;
+
+        public ReadyToRunXmlRootProvider(Stream documentStream, ManifestResource resource, ModuleDesc owningModule, string xmlDocumentLocation)
+        {
+            _context = owningModule.Context;
+            _documentStream = documentStream;
+            _resource = resource;
+            _owningModule = owningModule;
+            _xmlDocumentLocation = xmlDocumentLocation;
+        }
+
+        public void AddCompilationRoots(IRootingServiceProvider rootProvider)
+        {
+            CompilationRootProvider root = new CompilationRootProvider(rootProvider, _context, _documentStream, _resource, _owningModule, _xmlDocumentLocation);
+            root.ProcessXml();
+        }
+
+        public static bool TryCreateRootProviderFromEmbeddedDescriptorFile(EcmaModule module, out ReadyToRunXmlRootProvider provider)
+        {
+            PEMemoryBlock resourceDirectory = module.PEReader.GetSectionData(module.PEReader.PEHeaders.CorHeader.ResourcesDirectory.RelativeVirtualAddress);
+
+            foreach (var resourceHandle in module.MetadataReader.ManifestResources)
+            {
+                ManifestResource resource = module.MetadataReader.GetManifestResource(resourceHandle);
+
+                // Don't try to process linked resources or resources in other assemblies
+                if (!resource.Implementation.IsNil)
+                {
+                    continue;
+                }
+
+                string resourceName = module.MetadataReader.GetString(resource.Name);
+                if (resourceName == "ILLink.Descriptors.xml")
+                {
+                    BlobReader reader = resourceDirectory.GetReader((int)resource.Offset, resourceDirectory.Length - (int)resource.Offset);
+                    int length = (int)reader.ReadUInt32();
+
+                    UnmanagedMemoryStream ms;
+                    unsafe
+                    {
+                        ms = new UnmanagedMemoryStream(reader.CurrentPointer, length);
+                    }
+
+                    provider = new ReadyToRunXmlRootProvider(ms, resource, module, "resource " + resourceName + " in " + module.ToString());
+                    return true;
+                }
+            }
+            provider = null;
+            return false;
+        }
+
+        private class CompilationRootProvider : ProcessLinkerXmlBase
+        {
+            private const string NamespaceElementName = "namespace";
+            private const string _required = "required";
+            private const string _preserve = "preserve";
+            private const string _accessors = "accessors";
+            private static readonly string[] _accessorsAll = new string[] { "all" };
+            private static readonly char[] _accessorsSep = new char[] { ';' };
```

Comment:
Some of these appear unreferenced

(It's one of the reasons I'm not a fan of introducing `const` for what is a public API that will never change - introducing const is always a maintenance cost that should be weighed against the benefit - but this is the IL Linker codebase style...).